packages/type/src/reflection/type.ts

```diff
@@ -1108,6 +1113,70 @@ export function indexAccess(container: Type, index: Type): Type {
         }
     } else if (container.kind === ReflectionKind.any) {
         return { kind: ReflectionKind.any };
+    } else if (container.kind === ReflectionKind.union) {
+        if (index.kind === ReflectionKind.literal) {
+            // Deals with indexing a union with a literal.
+            // For example, if you have a union of {foo: 'bar'} | {foo: 'baz'}
+            // and you index it with 'foo', you get 'bar' | 'baz'. This should
+            // accordingly print ['bar', 'baz'] when valueOf<...>() is called
+            // on the union.
+            if (['string', 'number', 'symbol'].includes(typeof index.literal)) {
+                const union: TypeUnion = { kind: ReflectionKind.union, types: [] };
+
+                // For each type in the union, t, resolve the type at index.
+                for (const t of container.types) {
+                    const resolvedType = indexAccess(t, index);
```

Comment:
This version uses recursion and covers many more cases, more thoroughly. Such as indexing tuples as well.

I do get the impression that you were avoiding recursion, though, so if you would prefer without see the previous commit - which I had working with objects and classes.